load-testing/initial-load-testing.md

```diff
@@ -0,0 +1,89 @@
+
+<sup>5</sup> The majority of backend checks that did pass should have completed a lot faster because each check for an individual file should only take around 1 minute. This suggests that SQS is sending a lot of requests to each individual lambda which is increasing the time it takes.
+
+<sup>6</sup> The browser crashed while trying to send the client side processing results to the API. The response json from the client metadata call is around 50Mb so the browser is running out of memory. I tried this in Chrome and Firefox. I did get the upload to succeed by using a t3.2xlarge which has 8 cores and 32Gb of RAM but this is much larger than the expected machines the users will use though.
+
+## Problems found during the tests
+
+### Timeouts
+There are 4 timeouts that were preventing large consignments being processed.
+* The akka request timeout. This is the maximum time a request can take
+* The akka server timeout. The maximum time a server response can take. This needs to be the same as the request timeout as they basically do the same thing.
+* The load balancer idle timout.
+* The client timeout when contacting the API. When the API is overloaded, requests can take longer and the clients are timing out.
+
+### Logging
+There is little to no logging in the backend checks lambdas. The antivirus prints the current file id but the others only log errors. This makes it difficult to find out what has happened when there are a lot of files processed. Also, we get one line of a stack trace per cloudwatch event which makes them difficult to read.
+
+### Client metadata api response json
+We are returning a large json payload from the client metadata mutation which is consuming a lot of memory in the browser and causing the browser to crash on lower spec machines.
+
+### No mulipart uploads in export task
+We are using the single file upload call in the export task which has a size limit of 5Gb
+
+### Parallelism in the backend check lambdas
+It seems that AWS is passing a lot of messages to individual lambdas which is making them slow. We can reduce the number of messages passed to an individual lambda but this will increase the cost as more lambdas will run.
+
+### Too many requests to the API
+The API update lambda is making a single request for every file which is partly why the API runs so slowly under load. 
+
+### API being overloaded
+The API was responding very slowly under load although the CPU and memory usage never moved much above 50%. This causes numerous problems inculding clients timing out and the web app itself running slowly for other users.
+
+### Lambda using all available memory
+The checksum and antivirus lambdas use all their available memory for larger files. Increasing it may improve the speed of the check but we'd have to run some more tests to be sure.
+
+## Potential fixes
+I'm going to group these roughly into quick fixes, medium effort fixes and more difficult changes so we can see what we want to prioritise.
```

Comment:
I was expecting pagination to come up here, because I thought the bottleneck on large consignments was one of the JavaScript API requests for creating files or saving the client-side metadata. Is that what happened in practice?

Reply:
That is what happened yeah and that was mostly what caused the API to slow down so much. I've added in a fix to add pagination. I've added it into the medium effort fixes just because it'll take a bit of time to work out which page sizes work best.